test-suite/test-functionality.sh

```diff
@@ -0,0 +1,264 @@
+echo "TMPDIR=${TMPDIR:=${RUNNER_TEMP:-/tmp}}"
+# * directory for cloning git repositories containing various test tools
+echo "CLONES_DIR=${CLONES_DIR:=$TMPDIR/clones}"
+# * directories of cloned repositories
+echo "DAFT_DIR=${DAFT_DIR:=$CLONES_DIR/daft}"
+echo "SQUID_DAFTS_DIR=${SQUID_DAFTS_DIR:=$CLONES_DIR/squid-dafts}"
+echo "SQUID_OVERLORD_DIR=${SQUID_OVERLORD_DIR:=$CLONES_DIR/squid-overlord}"
+
+# print an error message (with special markers recognized by Github Actions)
+echo_error() {
+    echo "::error ::" "$@"
+}
+
+# print a warning message (with special markers recognized by Github Actions)
+echo_warning() {
+    echo "::warning ::" "$@"
+}
+
+run() {
+    echo "running: $@"
+    "$@"
+}
+
+# Whether the branch has a commit with a message matching the given regex.
+# The first argument is the SHA of the primary commit with the desired change.
+# We want to find up/backported commits that will have different SHAs and that
+# may have minor commit message variations, so that SHA is for reference only.
+has_commit_by_message() {
+    local commit="$1"
+    shift
+
+    if git log --grep "$@" | grep -q ^commit
+    then
+        return 0;
+    fi
+
+    echo "Was looking for branch commit with a message matching the following"
+    echo "    regex: " "$@"
+    echo "    This code lacks commit $commit or equivalent."
+    return 1;
+}
+
+clone_repo() {
+    local repo_url="$1"
+    local destination_dir="$2"
+
+    if test -e $destination_dir
+    then
+        echo "Skipping already fetched $destination_dir"
+    elif run git clone --no-tags --quiet --depth=1 --branch production -- "$repo_url" "$destination_dir"
+    then
+        if test -e "$destination_dir/package.json"
+        then
+            ( cd $destination_dir && run npm install --no-audit --no-save ) || return
+        fi
+    else
+        echo_error "Failed to fetch $repo_url into $destination_dir"
+        return 1
+    fi
+
+    (cd $destination_dir && echo "Using commit `git rev-parse HEAD`")
+}
+
+start_overlord() {
+    local url=http://localhost:13128/
+    local log=$TMPDIR/squid-overlord.log
+    if test -e $log && curl -H 'Pop-Version: 4' --no-progress-meter $url/check > /dev/null
+    then
+        echo "Will reuse squid-overlord service running at $url"
+        return 0;
+    fi
+
+    # Do not be tempted to simply run `sudo ... overlord.pl`: User nobody will
+    # lack read permissions, and sudo will ask for a password.
+    sudo -n --background -u nobody perl < $SQUID_OVERLORD_DIR/overlord.pl > $log 2>&1 || return
+    echo "Started squid-overlord service at $url"
+}
+
+setup_test_tools() {
+    echo "::group::Setup test tools"
+
+    clone_repo https://github.com/measurement-factory/daft $DAFT_DIR || return
+    clone_repo https://github.com/measurement-factory/squid-dafts $SQUID_DAFTS_DIR || return
+    clone_repo https://github.com/measurement-factory/squid-overlord $SQUID_OVERLORD_DIR || return
```

Comment:
The above three repositories contain Factory test cases and the tools they are based on. I do not like the fact that we are hard-coding such a dependency into Squid repository, but I see no better way. This is one of the negative side effects of using invasive CI configurations that dominate CI tools these days...

One possible alternative is to move these scripts into our squid-ci repository and download them here. However, that feels like swimming against the current again: If we are going to use invasive configurations, then we should not essentially move big parts of those configurations into a different repository, recreating problems that invasive configurations are meant to solve/avoid!

Reply:
IMO moving those script to the squid-ci repository leaves the system in a better overall state. Especially since we (the full community) can then review and add to the test set. But agree that is not a good change for this PR. I recommend doing it as a followup though.

Reply:
Thank you. Things are indeed clearer. I am in agreement with the current implementation going in as-is.